### PR TITLE
fetchFromGitHub add LFS support & Spacenav updates

### LINKED ISSUE #230295: Spacenav module does not work in Blender
### Describe the bug
With the spacenavd module active, the 3D mouse is not working.

### Steps To Reproduce
1. Set `hardware.spacenavd.enable = true`
2. Open Blender
3. The 3D Mouse is not working

### Expected behavior
With the module is active, the mouse should be detected and working in Blender

### Additional context
If you start the binary `spacenavd` with `sudo` seems to be enough for Blender to be enough for it to work.

When the blender binary detects the mouse the first line is 
`sh: line 1: lsusb : command not found` weirdly

### Notify maintainers
@Sohalt 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
- system: `"x86_64-linux"`
 - host os: `Linux 6.1.15, NixOS, 23.05 (Stoat), 23.05pre-git`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.14.1`
 - channels(root): `"nix-ld, nixos"`
 - channels(sirc): `""`
 - nixpkgs: `/nix/store/vnymyf1cja5h41zlg9bxg6a9sq5chgs5-nixos-unstable-2023-03-06`
```

### PR DESCRIPTION
**Pull request by Sohalt**
## Description of changes

- Update libspnav
- Update spacenavd
- Support LFS in fetchFromGitHub
- Update spnavcfg
- Update spacenavd module

fixes https://github.com/NixOS/nixpkgs/issues/230295

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by the-furry-hubofeverything** [comment]
IMO adding code to a fetcher is a fairly large change, since it is used by everyone. I suggest either add that to the title or create a seperate PR

File: nixos/modules/services/hardware/spacenavd.nix [hunk 1]
```diff
@@ -13,12 +13,14 @@ in {
   };
 
   config = mkIf cfg.enable {
-    systemd.user.services.spacenavd = {
+    systemd.user.services.spacenavd = mkIf (!versionAtLeast pkgs.spacenavd.version "1.1") {
```
Comment (jue89) [inline nixos/modules/services/hardware/spacenavd.nix:16]: I'm not a NixOS expert, but when you update `pkgs.spacenavd` you could skip this fallback. Or what's the idea behind this fallback?

Comment (Sohalt) [reply nixos/modules/services/hardware/spacenavd.nix:16]: The idea is to be able to still use the module with old versions of spacenavd, e.g. if you have an overlay with a custom, patched spacenavd. Might be overkill though.

Comment (Sohalt) [reply nixos/modules/services/hardware/spacenavd.nix]: I removed the fallback. People updating to the new module will probably also use the new package.

**Comment by Sohalt** [comment]
Tested the module and spnavcfg. Also tested that it works in Blender 3.5.1 (using a spacemouse wireless)

**Comment by Sohalt** [comment]
fixes  #230295

**Comment by luzpaz** [comment]
What is blocking this PR ?

**Comment by coldelectrons** [comment]
I am also interested in this PR's progress.

**Comment by AT-Geist** [comment]
I would also like to see this merged

File: pkgs/misc/drivers/spacenavd/default.nix [hunk 2]
```diff
@@ -1,33 +1,16 @@
 { stdenv, lib, fetchFromGitHub, fetchpatch, libX11, IOKit }:
```
Comment (SuperSandro2000) [inline pkgs/misc/drivers/spacenavd/default.nix]:
```suggestion
{ stdenv, lib, fetchFromGitHub, libX11, IOKit }:
```

File: pkgs/misc/drivers/spacenavd/default.nix [hunk 3]
```diff
@@ -18,6 +18,12 @@ stdenv.mkDerivation rec {
 
   makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
 
+  postInstall = ''
+    install -Dm644 $src/contrib/systemd/spacenavd.service -t $out/lib/systemd/system
+    substituteInPlace $out/lib/systemd/system/spacenavd.service \
+      --replace "/usr/local/bin/spacenavd" "$out/bin/spacenavd"
```
Comment (SuperSandro2000) [inline pkgs/misc/drivers/spacenavd/default.nix:34]:
```suggestion
      --replace-fail "/usr/local/bin/spacenavd" "$out/bin/spacenavd"
```

File: pkgs/applications/misc/spnavcfg/default.nix [hunk 4]
```diff
@@ -1,33 +1,28 @@
...
-  version = "0.3.1";
+  version = "1.1";
 
   src = fetchFromGitHub {
     owner = "FreeSpacenav";
     repo = pname;
     rev = "v${version}";
-    sha256 = "180mkdis15gxs79rr3f7hpwa1p6v81bybw37pzzdjnmqwqrc08a0";
+    sha256 = "sha256-P3JYhZnaCxzJETwC4g5m4xAGBk28/Va7Z/ybqwacIaA=";
+    fetchLFS = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/spnavcfg/default.nix:12]:
```suggestion
    fetchLFS = true;
    sha256 = "sha256-P3JYhZnaCxzJETwC4g5m4xAGBk28/Va7Z/ybqwacIaA=";
```

File: pkgs/applications/misc/spnavcfg/default.nix [hunk 5]
```diff
@@ -1,33 +1,28 @@
...
-    sed -i s/4775/775/ Makefile.in
-  '';
+  nativeBuildInputs = [ pkg-config wrapQtAppsHook ];
+  buildInputs = [ qtbase libspnav ];
 
-  nativeBuildInputs = [ pkg-config ];
-  buildInputs = [ gtk2 ];
-
-  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
+  #makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/spnavcfg/default.nix:25]: Please drop the comment

File: nixos/modules/services/hardware/spacenavd.nix [hunk 6]
```diff
@@ -13,12 +13,7 @@ in {
...
   config = mkIf cfg.enable {
-    systemd.user.services.spacenavd = {
-      description = "Daemon for the Spacenavigator 6DOF mice by 3Dconnexion";
-      wantedBy = [ "graphical.target" ];
-      serviceConfig = {
-        ExecStart = "${pkgs.spacenavd}/bin/spacenavd -d -l syslog";
-      };
-    };
+    systemd.packages = [pkgs.spacenavd];
+    systemd.services.spacenavd.enable = true;
```
Comment (SuperSandro2000) [inline nixos/modules/services/hardware/spacenavd.nix:17]:
```suggestion
    systemd = {
      packages = [ pkgs.spacenavd ];
      services.spacenavd.enable = true;
    };
```

File: pkgs/misc/drivers/spacenavd/default.nix [hunk 7]
```diff
@@ -1,22 +1,32 @@
...
+      libX11
+      libXext
+    ]
     ++ lib.optional stdenv.isDarwin IOKit;
 
-  configureFlags = [ "--disable-debug" ];
+  configureFlags = ["--disable-debug"];
 
-  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
+  makeFlags = ["CC=${stdenv.cc.targetPrefix}cc"];
```
Comment (SuperSandro2000) [inline pkgs/misc/drivers/spacenavd/default.nix:29]: That's not nixfmt-rfc-style, is it?

File: pkgs/misc/drivers/spacenavd/default.nix [hunk 8]
```diff
@@ -1,22 +1,32 @@
...
-  version = "1.2";
+  version = "1.3";
   pname = "spacenavd";
 
   src = fetchFromGitHub {
     owner = "FreeSpacenav";
     repo = "spacenavd";
     rev = "v${version}";
-    sha256 = "sha256-UuM/HTgictvIvlUnHZ5Ha8XwBhDTbt7CG9c4jzKQl0s=";
+    sha256 = "sha256-26geQYOXjMZZ/FpPpav7zfql0davTBwB4Ir+X1oep9Q=";
```
Comment (SuperSandro2000) [inline pkgs/misc/drivers/spacenavd/default.nix:17]: This should be hash

File: pkgs/top-level/all-packages.nix [hunk 9]
```diff
@@ -38848,7 +38848,7 @@ with pkgs;
 
   spacenav-cube-example = callPackage ../applications/misc/spacenav-cube-example { };
 
-  spnavcfg = callPackage ../applications/misc/spnavcfg { };
+  spnavcfg = libsForQt5.callPackage ../applications/misc/spnavcfg { };
```
Comment (philiptaron) [inline pkgs/top-level/all-packages.nix:38851]: As I commented on the `spnavcfg` package, I'd much rather that the package contain the details about which version of qt is needed than `all-packages.nix`.

File: pkgs/misc/drivers/spacenavd/default.nix [hunk 10]
```diff
@@ -1,40 +1,37 @@
...
 
   configureFlags = [ "--disable-debug" ];
 
   makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
 
+  postInstall = ''
+    install -Dm644 $src/contrib/systemd/spacenavd.service -t $out/lib/systemd/system
+    substituteInPlace $out/lib/systemd/system/spacenavd.service \
+      --replace-fail "/usr/local/bin/spacenavd" "$out/bin/spacenavd"
+  '';
```
Comment (philiptaron) [inline pkgs/misc/drivers/spacenavd/default.nix:33]: As I understand it, this moves the systemd service from the NixOS configuration module to this package. Is that right?

File: pkgs/applications/misc/spnavcfg/default.nix [hunk 11]
```diff
@@ -1,33 +1,40 @@
-{ stdenv, lib, fetchFromGitHub, pkg-config, gtk2 }:
-
+{
+  stdenv,
+  lib,
+  fetchFromGitHub,
+  fetchpatch,
+  pkg-config,
+  libspnav,
+  qtbase,
```
Comment (philiptaron) [inline pkgs/applications/misc/spnavcfg/default.nix:8]: I'd prefer if this needs qt5 to have that be encoded in the package, not in `all-packages.nix`.

**Review by philiptaron — DISMISSED** [review]
One change requested: `spnavcfg` ought to contain the details of which version of qt it depends on, not `all-packages.nix`.

I'm going to merge once that's resolved.

The change to `fetchGitHub`, which is what got me added to this PR, looks great. Thanks.

The three packages that fail to build also fail to build on master.

Result of `nixpkgs-review pr 263120` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>horizon-eda</li>
    <li>super-slicer</li>
    <li>super-slicer.debug</li>
  </ul>
</details>
<details>
  <summary>27 packages built:</summary>
  <ul>
    <li>blender</li>
    <li>blender-hip</li>
    <li>blendfarm</li>
    <li>cameractrls</li>
    <li>cameractrls-gtk3</li>
    <li>cameractrls-gtk4</li>
    <li>disko</li>
    <li>dune3d</li>
    <li>freecad</li>
    <li>freecad-wayland</li>
    <li>kicadAddons.kikit</li>
    <li>kicadAddons.kikit-library</li>
    <li>kikit</li>
    <li>kikit.dist</li>
    <li>libspnav</li>
    <li>openscad</li>
    <li>openscad-unstable</li>
    <li>snis</li>
    <li>solvespace</li>
    <li>spacenav-cube-example</li>
    <li>spacenavd</li>
    <li>spnavcfg</li>
    <li>super-slicer-beta</li>
    <li>super-slicer-beta.debug</li>
    <li>super-slicer-latest</li>
    <li>super-slicer-latest.debug</li>
    <li>vimPlugins.openscad-nvim</li>
  </ul>
</details>

**Comment by jue89** [comment]
> The three packages that fail to build also fail to build on master.

FYI ... `horizon-eda` should be fixed by https://github.com/NixOS/nixpkgs/pull/339213.

**Comment by philiptaron** [comment]
> > The three packages that fail to build also fail to build on master.
> 
> FYI ... `horizon-eda` should be fixed by #339213.

Thanks, merged.

**Comment by philiptaron** [comment]
I'd like to merge this. @Sohalt, would you like to me make the adjustment that I requested?

File: pkgs/development/libraries/libspnav/default.nix [hunk 12]
```diff
@@ -1,25 +1,19 @@
 { stdenv, lib, fetchFromGitHub, libX11, fixDarwinDylibNames }:
 
 stdenv.mkDerivation rec {
-  version = "0.2.3";
+  version = "1.1";
   pname = "libspnav";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libspnav/default.nix:5]:
```suggestion
  pname = "libspnav";
  version = "1.1";
```

File: pkgs/development/libraries/libspnav/default.nix [hunk 13]
```diff
@@ -1,25 +1,19 @@
...
 
   src = fetchFromGitHub {
     owner = "FreeSpacenav";
     repo = "libspnav";
-    rev = "${pname}-${version}";
-    sha256 = "098h1jhlj87axpza5zgy58prp0zn94wyrbch6x0s7q4mzh7dc8ba";
+    rev = "v${version}";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libspnav/default.nix:10]:
```suggestion
    rev = "refs/tags/v${version}";
```

File: pkgs/misc/drivers/spacenavd/default.nix [hunk 14]
```diff
@@ -1,40 +1,37 @@
...
+  lib,
+  fetchFromGitHub,
+  libXext,
+  libX11,
+  IOKit,
+}:
 stdenv.mkDerivation rec {
-  version = "0.8";
+  version = "1.3";
   pname = "spacenavd";
```
Comment (SuperSandro2000) [inline pkgs/misc/drivers/spacenavd/default.nix:11]:
```suggestion
  pname = "spacenavd";
  version = "1.3";
```

File: pkgs/misc/drivers/spacenavd/default.nix (hunk 10 above)
Comment (SuperSandro2000) [reply pkgs/misc/drivers/spacenavd/default.nix:33]: The upstream service file is now used. Before we had the bare minimum service that starts plus a description.

**Comment by jue89** [comment]
> I'd like to merge this. @Sohalt, would you like me to make the adjustment that I requested? I'd be happy to do so.

@philiptaron FYI - @Sohalt gave you a thumbs-up on that comment ;o)

**Comment by philiptaron** [comment]
Missed that. Thanks! I'll do so when I have access to Big Computer not just Tiny Computer.

**Comment by jue89** [comment]
Gentle ping :ping_pong: 

I guess it's too late for 24.11?

**Comment by philiptaron** [comment]
I'll address @SuperSandro2000's requested changes, wait for ofBorg, then merge. Not too late for 24.11, @jue89 -- and thanks for the ping.

**Comment by philiptaron** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 263120`

---
### `aarch64-darwin`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>spnavcfg</li>
    <li>super-slicer</li>
    <li>super-slicer-beta</li>
    <li>super-slicer-latest</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>dune3d</li>
    <li>libspnav</li>
    <li>spacenav-cube-example</li>
    <li>spacenavd</li>
  </ul>
</details>

**Comment by philiptaron** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 263120`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>super-slicer</li>
    <li>super-slicer.debug</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 27 packages built:</summary>
  <ul>
    <li>blender</li>
    <li>blender-hip</li>
    <li>blendfarm</li>
    <li>cameractrls</li>
    <li>cameractrls-gtk3</li>
    <li>cameractrls-gtk4</li>
    <li>dune3d</li>
    <li>freecad</li>
    <li>freecad-wayland</li>
    <li>horizon-eda</li>
    <li>kicadAddons.kikit</li>
    <li>kicadAddons.kikit-library</li>
    <li>kikit</li>
    <li>kikit.dist</li>
    <li>libspnav</li>
    <li>openscad</li>
    <li>openscad-unstable</li>
    <li>snis</li>
    <li>solvespace</li>
    <li>spacenav-cube-example</li>
    <li>spacenavd</li>
    <li>spnavcfg</li>
    <li>super-slicer-beta</li>
    <li>super-slicer-beta.debug</li>
    <li>super-slicer-latest</li>
    <li>super-slicer-latest.debug</li>
    <li>vimPlugins.openscad-nvim</li>
  </ul>
</details>

**Comment by philiptaron** [comment]
Need to fix the Darwin build failure, then I think it's good.

**Comment by philiptaron** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 263120`

---
### `aarch64-darwin`
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>super-slicer</li>
    <li>super-slicer-beta</li>
    <li>super-slicer-latest</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 5 packages built:</summary>
  <ul>
    <li>dune3d</li>
    <li>libspnav</li>
    <li>spacenav-cube-example</li>
    <li>spacenavd</li>
    <li>spnavcfg</li>
  </ul>
</details>

**Pull request merged**